src/main/java/io/nats/client/impl/NatsDispatcher.java

```diff
@@ -207,8 +275,32 @@ public Dispatcher unsubscribe(String subject, int after) {
         if (subject == null || subject.length() == 0) {
             throw new IllegalArgumentException("Subject is required in unsubscribe");
         }
-        
-        NatsSubscription sub = subscriptions.get(subject);
+
+        NatsSubscription sub = this.subscriptionsUsingDefaultHandler.get(subject);
+
+        if (sub != null) {
+            this.connection.unsubscribe(sub, after); // Connection will tell us when to remove from the map
+        }
+
+        return this;
+    }
+
+    public Dispatcher unsubscribe(Subscription subscription, int after) {
+        if (subscription.getDispatcher() != this) {
+            throw new IllegalStateException("Subscription is not managed by this Dispatcher");
+        }
+
+        if (this.subscriptionsUsingDefaultHandler.get(subscription.getSubject()) != null) {
```

Comment:
I don' think this is right. You could have "foo" under the default handler and "foo" under a custom handler. Instead you should check if the sid is in the with handler map and if that is false, then do this check.

Reply:
Derp! Great catch.

Reply:
I was trying to be clever here. We do end up checking against `subscriptionsWithHandlers` a few lines down (mimicking behavior of `subscribe(String, int)`), so I think we might actually be able to remove the entire conditional.

Reply:
the problem is if i subscribe("foo") and s = subscribe("foo", mh), then try to unsubscribe(s) this will throw an illegal state. But it should be ok. I think you can just remove the check.

Reply:
We should add a test around that example, because I think it could pose an issue in `Dispatcher#remove`. We should be fine if we do something like:
```java
    void remove(NatsSubscription sub) {
        if (this.subscriptionsWithHandlers.remove(sub.getSID()) != null) {
            this.subscriptionHandlers.remove(sub.getSID());
        } else {
            NatsSubscription s = this.subscriptionsUsingDefaultHandler.get(sub.getSubject());
            if (s.getSID() == sub.getSID()) {
                this.subscriptionsUsingDefaultHandler.remove(sub.getSubject());
            }
        }
    }
```
I'll push a commit up with comments if you're :+1: .

Reply:
+1